### ISSUE TITLE
Add non-interactive mode for `summoner`

### ISSUE THREAD

**Issue by willbasky** [kowainik/summoner#114]
I encountered such a problem. To test the changes in the template, I had create project each time and answer in interactive mode, although I knew the answers in advance.
Then I wrote down the answers in the `configuration.toml`, but some questions were still asked.

Is it possible to make `summoner` be quiet based on the configuration file?

**Comment by ChShersh** [comment]
@willbasky Could you tell what exact answers are asked?

**Comment by willbasky** [comment]
@ChShersh Fields in `conf.toml` are
```toml
cabal = true
stack = true

owner       = "willbasky"
description = "testAll"
fullName    = "willbasky"
email       = "email@gmail.com"
category    = "TestAll"
license     = "MIT"

github   = true
travis   = true
appveyor = false

lib      = true
exe      = true
test     = true
bench    = true

bscript  = false
private  = false
ghcVers  = ["8.2.2"]
extentions = []
```
Questions that are still asked by summoner:
- Repository owner:  [willbasky]
- Short project description: 
- Author:  [willbasky]
- Maintainer e-mail:  [email@gmail.com]
- List of categories to choose from: ...
- License: ...
- Add custom prelude? [y]/n
- Supported by 'summoner' GHCs: ...

It is less then without `conf.toml`, but they are.

**Comment by willbasky** [comment]
As i understand `summoner``s mission to simplify user`s experience on new project start. For the purpose interactive mode chosen. 
Idea of subject feature was born during testing process to attain more automation via `.toml`. It is feature for testers or advanced developers, non interactive mode.
It may be contradicts to main idea of summoner, i am not sure.

**Comment by ChShersh** [comment]
@vrom911 @willbasky `summoner` is not optimized for immediate project creation. Personally I think that projects shouldn't be created instantaneously  because this is the job that should be done only once and it's better to spend some time to double-check everything. `summoner` also pushes package directly to GitHub, so we want to decrease chances of user to screw up.

I agree with @vrom911 that this also a non-trivial implementation task. As I mentioned earlier, I don't want to have a lot of `if`s inside `summoner` because we already have a lot of fragile logic. But migration to proper structure involves **HUGE** refactoring and I would like to come up with design first instead of refactoring whole package :slightly_smiling_face: 

For example, we could introduce some flag like `--skip-checks` which chooses defaults and doesn't ask user. We won't need to refactor code much with this approach but this solution is very bad and doesn't scale.

I guess the problem you're trying to solve is to have an ability to test `summoner` faster. But it's better to implement some good testing system instead of some temporary workarounds.

Instead of skipping some checks we could think in different direction. For example, we could think on the process of _pressing <kbd>Enter</kbd> automatically_. Like, introduce some concept of **scenarios** where you have file like this:

```
<<< y # project
<<< n # cabal
<<< relude
<<<  # enter
```

This is much harder to implement but will also help to make testing automatic. Though, I can imagine that such task will consume whole month of some developer's time...

**Comment by willbasky** [comment]
`Add custom prelude? [y]/n`

> This can be done by introducing a bool in Config data type. Or we can change default here from Y to N so all you need to do is to press Enter. Needs to be discussed.

Often case is standard Prelude, so let we do it `n` by default. What do you think?

**Comment by vrom911** [comment]
After some discussions, I now think that it could be useful to some people to have such mode. Let's have an `--foo`(I don't know the suitable name) option which will:
1. Read the config file, if applicable
2. Read the command-line arguments, if applicable
3. Combine all kind of configurations as we do in the `new` command itself.
4. Create the project without asking any information, only showing what options are added/skipped.

I see a small issue. If you have `github = true` in the `.summoner.toml` file and don't specify either `--ignore-config` or `without -g` or `--offline` then this project will be uploaded to GitHub. Well, it's not an issue, it's correct behaviour, but I'm not sure, that everyone is happy with that.

**Comment by vrom911** [comment]
@ChShersh I agree that we need to make a function `Decision -> Bool` and we already done some work on this for the initial TUI interface. I guess this should help a lot, as we made a lot of this dependent checks and some decisions in there, but the thing is that there it all can be changed, so we should think more on what we did in there. I guess we can somehow share this functions with CLI and TUI.

The mentioned code is here:
https://github.com/kowainik/summoner/blob/b1cc85a78661eb23162da9034de5dea415b2210c/summoner-tui/src/Summoner/Tui/Kit.hs#L223-L291

**Comment by m-renaud** [comment]
My 2c :)

> Regarding name of the option

`-n, --non-interactive` is inline with `cabal init -n/--non-interactive`, so that's a +1 from me.

> Library and Executable are both set to Idk but we should choose at least one

I think Library _and_ Executable is a good default for a "batteries included, don't ask me things" project. From what I've seen and read, putting all of your logic into a library and having a "dumb `main`" is a very popular way of structuring code, and also makes testing the code easier since testing the modules included in an executable only is messier. This is obviously just an opinion :)

> Cabal and Stack are both set to Idk but we should choose at least one

I'm assuming if these were specified in the config file then those would be chosen, correct? If only one is specified, assume `false` for the other.

In the case when neither `stack=` nor `cabal=` are specified in `~/.summoner.toml` then its a bit trickier. One (relatively simple) option is to check to see which executables are on the `$PATH` and set the values based on that.

> Regarding --github

Yeah, this is tricky since this has an effect somewhere other than your local file system. Even if _non-interactive_ is requested I can still see an argument for double checking if `github = true`, but can be overridden by a `-f, --force` flag.

So, to summarize, we would have the following situations (assuming `-n` for example sake)

`github = false`, `summon new myapp -n`
=> Init, no prompts

`github = true`, `summon new myapp -n`
=> Init, prompts to double check regarding github upload

`github = true`, `summon new myapp -n -f
 => Init, no prompts

**Comment by ChShersh** [comment]
> I agree that we need to make a function Decision -> Bool and we already done some work on this for the initial TUI interface.

@vrom911 I agree with the fact that the interface should be unified to avoid code duplication but most importantly to keep this complicated logic in one place. Looks not that trivial but should be doable.

> I think Library and Executable is a good default for a "batteries included, don't ask me things" project. 

I'm also okay with having _Library and Executable_ enabled by default in this case.

> One (relatively simple) option is to check to see which executables are on the $PATH and set the values based on that.

This is indeed not a difficult solution according to this answer:

* https://stackoverflow.com/questions/47020908/programmatically-check-if-executable-exists-without-running-it-or-using-which

Alternatively, we can force users to specify either `--stack` or `--cabal` flag via CLI. Or, we can create just Cabal project and then write _info_ message like `Run 'stack init' command to initialize stack project`. Currently in our settings there's nothing that is excluded by enabling `stack`, enabling Stack only creates additional `stack.yaml` files and affects generated `.travis.yml`. But disabling `stack` also disables AppVeyor because we're currently supporting AppVeyor only with `stack`.

Your proposal regarding handling `--github` looks fine to me 👍
